bench/micro/Bench/Database/LSMTree/Internal/Merge.hs

```diff
@@ -0,0 +1,391 @@
+        { name         = "utxo-x4+1-max-skewed-lastlevel"  -- live levelling merge
+        , nentries     = scaled [1, 1, 1, 1, 16]
+        , finserts     = 1
+        , fdeletes     = 1
+        , mergeLevel   = Merge.LastLevel
+        }
+    ]
+  where
+    totalentries = 50_000 :: Word64
+
+    -- For each run, a significant part of all possible keys are in use, e.g.
+    -- half of them for 4 runs of equal size. This leads to many collisions.
+    -- Note that each input run still has the full number of elements due to the
+    -- use of 'sampleWithoutReplacement'.
+    -- Warning: Using this with a run of around half of all entries (or more)
+    -- leads  to issues with generating distinct keys for all its entries!
+    randomWord64WithCollisions :: Rnd SerialisedKey
+    randomWord64WithCollisions =
+          first (serialiseKey . Hash.hash64)
+        . uniformR (0, totalentries `div` 2)
+
+    scaled weights =
+      let total = sum weights
+      in [ round (fromIntegral totalentries * w / total :: Double)
+         | w <- weights
+         ]
+
+benchMerge :: Config -> Benchmark
+benchMerge conf@Config{name} =
+    withEnv $ \ ~(_dir, hasFS, runs) ->
+      bgroup name [
+          bench "merge" $
+            -- We'd like to do: `whnfAppIO (runs' -> ...) runs`.
+            -- However, we also need per-run cleanup to avoid running out of
+            -- disk space. We use `perRunEnvWithCleanup`, which has two issues:
+            -- 1. Just as `whnfAppIO` etc., it takes an IO action and returns
+            --    `Benchmarkable`, which does not compose. As a workaround, we
+            --    thread `runs` through the environment, too.
+            -- 2. It forces the result to normal form, which would traverse the
+            --    whole run, so we force to WHNF ourselves and just return `()`.
+            Cr.perRunEnvWithCleanup
+              ((,) runs <$> getPaths)
+              (cleanupPaths hasFS . snd) $ \(runs', p) -> do
+                !run <- merge hasFS conf p runs'
+                -- Make sure to immediately close resulting runs so we don't run
+                -- out of file handles. Ideally this would not be measured, but at
+                -- least it's pretty cheap.
+                Run.removeReference hasFS run
+        ]
+  where
+    withEnv =
+        Cr.envWithCleanup
+          (writeBufferEnv conf)
+          writeBufferEnvCleanup
+
+    -- We'll remove the files on every run, so we can re-use the same run number.
+    getPaths :: IO Run.RunFsPaths
+    getPaths = pure outputRunNumber
+
+    -- We need to keep the other runs, but remove the freshly created one.
+    cleanupPaths :: FS.HasFS IO FS.HandleIO -> Run.RunFsPaths -> IO ()
+    cleanupPaths hasFS paths = do
+        traverse_ (FS.removeFile hasFS) (Run.runFsPaths paths)
+        exists <- FS.doesFileExist hasFS (Run.runChecksumsPath paths)
+        when exists $
+          FS.removeFile hasFS (Run.runChecksumsPath paths)
```

Comment:
Maybe rename these functions so it's clear that they are only supposed to touch the output run

Reply:
I simplified that part a bit and called the function `removeOutputRunFiles`.